### DISCUSSION TITLE
OPT-111-Lot-FiscalLegis not mandatory in SDK but mandatory in XSD - notice 29

### DISCUSSION THREAD

**Discussion by PiotrRymarz** [OP-TED/eForms-SDK#903] [Q&A] [answered]
Hi,

I am currently creating notice 29. Field OPT-111-Lot-FiscalLegis is present there. Could you help me understand why it is not marked as mandatory in SDK, although XSD requires it during validation? SDK version 1.10.1

OPT-111-Lot-FiscalLegis
from fields.json
`{
    "id" : "OPT-111-Lot-FiscalLegis",
    "parentNodeId" : "ND-LotFiscalLegislation",
    "name" : "Fiscal Legislation Document ID",
    "btId" : "OPT-111",
    "xpathAbsolute" : "/*/cac:ProcurementProjectLot[cbc:ID/@schemeName='Lot']/cac:TenderingTerms/cac:FiscalLegislationDocumentReference/cbc:ID",
    "xpathRelative" : "cbc:ID",
    "xsdSequenceOrder" : [ { "cbc:ID" : 2 } ],
    "type" : "id",
    "repeatable" : {
      "value" : false,
      "severity" : "ERROR"
    },
    "forbidden" : {
      "value" : false,
      "severity" : "ERROR",
      "constraints" : [ {
        "noticeTypes" : [ "1", "2", "3", "4", "5", "6", "CEI", "T01", "T02", "X01", "X02" ],
        "value" : true,
        "severity" : "ERROR"
      } ]
    }
  }`
  
  As you can see in notice 29 it is neither forbidden nor mandatory. But when it comes to validation in this field is mandatory and whole validation fails.
  
  `   <xsd:complexType name="DocumentReferenceType">
      <xsd:sequence>
        (...)
         <xsd:element ref="cbc:ID" minOccurs="1" maxOccurs="1"/>
       (...)
      </xsd:sequence>`
      
 Here is generated xml:
 `<?xml version="1.0" encoding="UTF-8"?>
<ContractAwardNotice xmlns="urn:oasis:names:specification:ubl:schema:xsd:ContractAwardNotice-2"
                     xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
                     xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"
                     xmlns:efac="http://data.europa.eu/p27/eforms-ubl-extension-aggregate-components/1"
                     xmlns:efbc="http://data.europa.eu/p27/eforms-ubl-extension-basic-components/1"
                     xmlns:efext="http://data.europa.eu/p27/eforms-ubl-extensions/1"
                     xmlns:ext="urn:oasis:names:specification:ubl:schema:xsd:CommonExtensionComponents-2">
  <ext:UBLExtensions>
    <ext:UBLExtension>
      <ext:ExtensionContent>
        <efext:EformsExtension>
          <efac:NoticeResult>
            <efac:LotResult>
              <cbc:ID schemeName="result">RES-0001</cbc:ID>
              <cbc:TenderResultCode listName="winner-selection-status">selec-w</cbc:TenderResultCode>
              <efac:LotTender>
                <cbc:ID schemeName="tender">TEN-0001</cbc:ID>
              </efac:LotTender>
              <efac:SettledContract>
                <cbc:ID schemeName="contract">CON-0001</cbc:ID>
              </efac:SettledContract>
              <efac:TenderLot>
                <cbc:ID schemeName="Lot">LOT-0001</cbc:ID>
              </efac:TenderLot>
            </efac:LotResult>
            <efac:LotTender>
              <cbc:ID schemeName="tender">TEN-0001</cbc:ID>
              <efac:TenderingParty>
                <cbc:ID schemeName="tendering-party">TPA-0001</cbc:ID>
              </efac:TenderingParty>
              <efac:TenderLot>
                <cbc:ID schemeName="Lot">LOT-0001</cbc:ID>
              </efac:TenderLot>
              <efac:TenderReference>
                <cbc:ID>dffgdfg</cbc:ID>
              </efac:TenderReference>
            </efac:LotTender>
            <efac:SettledContract>
              <cbc:ID schemeName="contract">CON-0001</cbc:ID>
              <cbc:AwardDate>2024-04-09Z</cbc:AwardDate>
              <cbc:IssueDate>2024-04-11Z</cbc:IssueDate>
              <efac:ContractReference>
                <cbc:ID schemeName="contract">rvdhdvdhv</cbc:ID>
              </efac:ContractReference>
              <efac:LotTender>
                <cbc:ID schemeName="tender">TEN-0001</cbc:ID>
              </efac:LotTender>
            </efac:SettledContract>
            <efac:TenderingParty>
              <cbc:ID schemeName="tendering-party">TPA-0001</cbc:ID>
            </efac:TenderingParty>
          </efac:NoticeResult>
          <efac:NoticeSubType>
            <cbc:SubTypeCode listName="notice-subtype">29</cbc:SubTypeCode>
          </efac:NoticeSubType>
          <efac:Organizations>
            <efac:Organization>
              <efac:Company>
                <cbc:WebsiteURI>https://dev.ezamowienia.gov.pl/emo-client/notices/284b5042-b6ad-4742-9f97-997c3679aa23</cbc:WebsiteURI>
                <cbc:EndpointID>https://dev.ezamowienia.gov.pl/emo-client/notices/284b5042-b6ad-4742-9f97-997c3679aa23</cbc:EndpointID>
                <cac:PartyIdentification>
                  <cbc:ID schemeName="organization">ORG-0001</cbc:ID>
                </cac:PartyIdentification>
                <cac:PartyName>
                  <cbc:Name languageID="POL">org 1</cbc:Name>
                </cac:PartyName>
                <cac:PostalAddress>
                  <cbc:Department>hr</cbc:Department>
                  <cbc:CityName>warszawa</cbc:CityName>
                  <cbc:PostalZone>01-123</cbc:PostalZone>
                  <cbc:CountrySubentityCode listName="nuts">PL911</cbc:CountrySubentityCode>
                  <cac:Country>
                    <cbc:IdentificationCode listName="country">POL</cbc:IdentificationCode>
                  </cac:Country>
                </cac:PostalAddress>
                <cac:PartyLegalEntity>
                  <cbc:CompanyID>nr rej</cbc:CompanyID>
                </cac:PartyLegalEntity>
                <cac:Contact>
                  <cbc:Telephone>123123123</cbc:Telephone>
                  <cbc:ElectronicMail>a2@com.pl</cbc:ElectronicMail>
                </cac:Contact>
              </efac:Company>
              <efac:TouchPoint>
                <cac:PartyIdentification>
                  <cbc:ID schemeName="touchpoint">TPO-0001</cbc:ID>
                </cac:PartyIdentification>
              </efac:TouchPoint>
            </efac:Organization>
          </efac:Organizations>
        </efext:EformsExtension>
      </ext:ExtensionContent>
    </ext:UBLExtension>
  </ext:UBLExtensions>
  <cbc:UBLVersionID>2.3</cbc:UBLVersionID>
  <cbc:CustomizationID>eforms-sdk-1.10</cbc:CustomizationID>
  <cbc:ID schemeName="notice-id">284b5042-b6ad-4742-9f97-997c3679aa23</cbc:ID>
  <cbc:ContractFolderID>ocds-148610-f27287d1-d9ee-11ee-a5a8-b2b724e7c200</cbc:ContractFolderID>
  <cbc:IssueDate>2024-04-13Z</cbc:IssueDate>
  <cbc:IssueTime>10:25:39Z</cbc:IssueTime>
  <cbc:VersionID>01</cbc:VersionID>
  <cbc:RequestedPublicationDate>2024-04-13Z</cbc:RequestedPublicationDate>
  <cbc:RegulatoryDomain>32014L0025</cbc:RegulatoryDomain>
  <cbc:NoticeTypeCode listName="result">can-standard</cbc:NoticeTypeCode>
  <cbc:NoticeLanguageCode>POL</cbc:NoticeLanguageCode>
  <cac:ContractingParty>
    <cac:ContractingPartyType>
      <cbc:PartyTypeCode listName="buyer-legal-type">grp-p-aut</cbc:PartyTypeCode>
    </cac:ContractingPartyType>
    <cac:ContractingActivity>
      <cbc:ActivityTypeCode listName="authority-activity">education</cbc:ActivityTypeCode>
    </cac:ContractingActivity>
    <cac:Party>
      <cac:PartyIdentification>
        <cbc:ID schemeName="organization">ORG-0001</cbc:ID>
      </cac:PartyIdentification>
    </cac:Party>
  </cac:ContractingParty>
  <cac:TenderingProcess>
    <cbc:ProcedureCode listName="procurement-procedure-type">open</cbc:ProcedureCode>
  </cac:TenderingProcess>
  <cac:ProcurementProject>
    <cbc:Name languageID="POL">dsfasdfa</cbc:Name>
    <cbc:Description languageID="POL">dsafdfsafsd</cbc:Description>
    <cbc:ProcurementTypeCode listName="contract-nature">services</cbc:ProcurementTypeCode>
    <cac:MainCommodityClassification>
      <cbc:ItemClassificationCode listName="cpv">71600000</cbc:ItemClassificationCode>
    </cac:MainCommodityClassification>
  </cac:ProcurementProject>
  <cac:ProcurementProjectLot>
    <cbc:ID schemeName="Lot">LOT-0001</cbc:ID>
    <cac:TenderingTerms>
      <cbc:FundingProgramCode listName="eu-funded">no-eu-funds</cbc:FundingProgramCode>
      <cac:FiscalLegislationDocumentReference>
        <cac:IssuerParty>
          <cac:PartyIdentification>
            <cbc:ID schemeName="touchpoint">TPO-0001</cbc:ID>
          </cac:PartyIdentification>
        </cac:IssuerParty>
      </cac:FiscalLegislationDocumentReference>
      <cac:AwardingTerms>
        <cac:AwardingCriterion>
          <cbc:CalculationExpression languageID="POL">opis</cbc:CalculationExpression>
          <cac:SubordinateAwardingCriterion>
            <cbc:AwardingCriterionTypeCode listName="award-criterion-type">price</cbc:AwardingCriterionTypeCode>
            <cbc:Name languageID="POL">fdsfds</cbc:Name>
            <cbc:Description languageID="POL">opis</cbc:Description>
          </cac:SubordinateAwardingCriterion>
        </cac:AwardingCriterion>
      </cac:AwardingTerms>
      <cac:AppealTerms>
        <cac:AppealReceiverParty>
          <cac:PartyIdentification>
            <cbc:ID schemeName="touchpoint">TPO-0001</cbc:ID>
          </cac:PartyIdentification>
        </cac:AppealReceiverParty>
      </cac:AppealTerms>
    </cac:TenderingTerms>
    <cac:TenderingProcess>
      <cbc:GovernmentAgreementConstraintIndicator>true</cbc:GovernmentAgreementConstraintIndicator>
      <cac:AuctionTerms>
        <cbc:AuctionConstraintIndicator>true</cbc:AuctionConstraintIndicator>
      </cac:AuctionTerms>
      <cac:ContractingSystem>
        <cbc:ContractingSystemTypeCode listName="framework-agreement">fa-w-rc</cbc:ContractingSystemTypeCode>
      </cac:ContractingSystem>
      <cac:ContractingSystem>
        <cbc:ContractingSystemTypeCode listName="dps-usage">none</cbc:ContractingSystemTypeCode>
      </cac:ContractingSystem>
    </cac:TenderingProcess>
    <cac:ProcurementProject>
      <cbc:ID schemeName="InternalID">id</cbc:ID>
      <cbc:Name languageID="POL">tytul</cbc:Name>
      <cbc:Description languageID="POL">opis</cbc:Description>
      <cbc:ProcurementTypeCode listName="contract-nature">services</cbc:ProcurementTypeCode>
      <cac:MainCommodityClassification>
        <cbc:ItemClassificationCode listName="cpv">71600000</cbc:ItemClassificationCode>
      </cac:MainCommodityClassification>
    </cac:ProcurementProject>
  </cac:ProcurementProjectLot>
  <cac:ProcurementProjectLot>
    <cbc:ID schemeName="LotsGroup">GLO-0001</cbc:ID>
  </cac:ProcurementProjectLot>
  <cac:TenderResult>
    <cbc:AwardDate>2000-01-01Z</cbc:AwardDate>
  </cac:TenderResult>
</ContractAwardNotice>
`

**Comment by YvesJo** [comment, marked as answer]
Hi,
This XML identifier is required by the XSD in the context of its parent element. It may not be marked as mandatory in the SDK as it would then require the presence of its parent element and other fields (OPT-110-Lot-FiscalLegis, OPT-301-Lot-FiscalLegis). this has to be addressed at XML generation; when further information about fiscal legislation is provided, then the identifier shall also exist, when no further information is provided, then the identifier should not be in the XML.

**Reply by PiotrRymarz** [reply]
Hi,
thank you for the answer.
Could you please help me understand your point. "It may not be marked as mandatory in the SDK as it would then require the presence of its parent element and other fields (OPT-110-Lot-FiscalLegis, OPT-301-Lot-FiscalLegis)." 
Let's look at OPT-140-Lot. It is also of type Id, it has also "xpathRelative" : "cbc:ID". It is forbidden conditionally based on it's parent and also mandatory in case of not being forbidden. How does OPT-140-Lot differ to  OPT-111-Lot-FiscalLegis so first one is marked as mandatory conditionally, and second one is not?

**Reply by YvesJo** [reply]
Hi,
The field OPT-111-Lot-FiscalLegis is mainly of interest to satisfy the schema when OPT-110-Lot-FiscalLegis or OPT-301-Lot-FiscalLegis exist, alone it has no added value and shouldn’t be present in the XML. Therefore, at XML generation, it should be checked first that further information (i.e. OPT-110-Lot-FiscalLegis or OPT-301-Lot-FiscalLegis) exist before generating the corresponding XML fragment.
Unconditioned Mandatory rules (Stage 2 Mandatory rules) require the XML element to exist without any condition and so does it for any XML node on its path; its XML parent node (cac:FiscalLegislationDocumentReference) is also the context that will trigger further rules for other fields of the set.
While Stage 2 Mandatory rule is not an option, let’s look at the option for Stage 4 Conditional rules.
There is an interdependency between these 3 fields which should be expressed somehow with rules. There are various options:
- Cover all possible dependencies and express all possible rules (with a number of rules varying with the square number of interdependent fields),
- Identify a field characteristic of the set of information and express the dependency of other fields from that field (with a number of rules that varies with the number of interdependent fields only).

The first option would lead to too many rules (causing performance issues) and possibly too many error messages for the user to deal with.
The second option has been privileged for the additional following reasons:
- OPT-111-Lot-FiscalLegis could be identified as the key field that characterizes this set of information, 
- OPT-111-Lot-FiscalLegis must be present when other fields of the group are.

Hence OPT-111-Lot-FiscalLegis may be identified as the “entry point” for that set of information and will condition the existence of the two other fields.
As long as the rules aren’t actually evaluated on the UI, choosing another field as “entry point” wouldn’t help because validation starts with the lexical validation (XSD conformance) before any further rules are applied, so in absence of OPT-111-Lot-FiscalLegis, the first error would come from the schema validation before any schematron rule is applied.
Having rules evaluated in the UI, would prevent the user to specify OPT-110-Lot-FiscalLegis or OPT-301-Lot-FiscalLegis when OPT-111-Lot-FiscalLegis is not itself specified (as both are conditionally forbidden in absence of OPT-111-Lot-FiscalLegis).
There are two options to consider (and may be more):
- Have a dynamic UI that integrates part of the rules set (displaying/hiding fields on the screen and highlighting the ones that become conditionally mandatory),
- Draw the author’s attention by highlighting the fields that must be filled in when information is provided.

I hope this helps clarifies my former comment. I may further develop on OPT-140-Lot when needed.
KR

**Reply by PiotrRymarz** [reply]
Thank you very much for this comprehensive answer.
